Initial commit for Azure Blob Storage adaptor

bensonmiller opened the pull request:
## Summary

Adds an OpenFn adaptor for Azure Blob Storage

## Details

This adaptor adds the following capabilities for working with Azure Blob Storage:

* Create a container
* Check the existence of a blob
* Fetch the properties of a blob
* Upload a blob
* Download a blob as JSON
* Download a blob as string

## Issues

This is my first adaptor PR and I'm not an especially competent JS developer. I'm certain there are issues. But I've subjected this adaptor to a fair amount of testing via the CLI and it is working respectably. Some thoughts:

* Your could make an argument that ALL azure-related functions should go in a single OpenFn adaptor. That's not what I've done here. This may be worth a discussion, but Blob Storage interactions /alone/ could grow to be a fairly substantial adaptor.
* I haven't spent any meaningful time trying to mock the Azure blob storage client(s), so the testing is embarrassing.
* I don't see a lot of try/catch logic in other adaptors, so I haven't put such logic in this one. Uncaught exceptions FTW!
* There are a number of different ways to authenticate against Azure. This adaptor only supports the `StorageSharedKeyCredential` at the moment.
* I wonder about the pattern of defining helper functions like `downloadAsJSON()` and `downloadAsString()`. The difference between these is minimal, but I saw a couple other adaptors doing similar things so I did it here. I'm not sure that I love it (or, even, that I've implemented it especially well).

## Review Checklist

Before merging, the reviewer should check the following items:

- [ ] Does the PR do what it claims to do?
- [ ] If this is a new adaptor, has the migration tool been run and the
      migration guide followed?
- [ ] Are there any unit tests? Should there be?
- [ ] Is there a changeset associated with this PR? Should there be? Note that
      dev only changes don't need a changeset.

bensonmiller commented:
Hello @josephjclark @stuartc. @taylordowns2000 suggested that I flag this to your attention. I suspect that this isn't yet in a merge-able state, but it feels close enough that I wanted to open a PR.

josephjclark commented:
@bensonmiller Thank you for this! Don't worry too much about the testing side - mocking is a really hard problem (we're working on a solution but no news yet).

I'll take a look at this on Monday and see if we can get it merged.

Does Azure have a sandbox that I can test this adaptor against? What's the effort for me, who thinks azure is a colour, to test this adaptor on my own machine?

bensonmiller commented:
@josephjclark Thank you! There are a couple ways to handle Azure testing. I could grant you access to a small Azure subscription that I have. That's not terrible, but would still require you to create and manage an account in Azure. Alternatively (and maybe the best approach) is to have you sign up for an Azure account. When you create an account for the first time, MSFT puts some cloud credits ($200 IIRC) into new accounts. Pretty standard stuff, but I can't imagine that would exceed those credits on blob storage testing in 1000 years. If you want to go down that second path, this is how I'd approach it:

1. Signup for an Azure account
2. Once you have an Azure subscription and are logged into the Azure web portal, create a resource group that is associated with the subscription. Call this resource group `openfnadaptortesting` or similar.
3. Create a storage account and assign it to that resource group. Call the storage account `blobstorageadaptortest` or similar. Accept all other defaults during the creation of the storage account, ensuring that the "Storage account key access" setting is enabled. You'll have a chance to confirm this setting on the review / summary screen immediately before pressing the "Create" button.
4. Once you've created the storage account, browse to the main "Storage Account" page (find it under the main navigation hamburger or via the search bar) and open the storage account that you just created. Then, in the left nav bar, find and click on the "Access Keys" item under "Security + Networking" heading.
5. Copy the Storage Account Name and the values of one of the keys (doesn't matter which key) into your `state.configuration` as the `accountName` and `accountKey` properties, respectively.

That should be all that's necessary to get a storage account that works with this adaptor. I think that's a relatively easy process and still (mostly) allows you to assert that Azure is just a color. I'm not far from that perspective myself, feeling much more comfortable in AWS and GCP. . .

josephjclark commented on packages/azureblobstorage/src/Adaptor.js:
```diff
@@ -0,0 +1,216 @@
+ * @param {object} uploadOptions - See BlockBlobUploadOptions in Azure Blob Storage docs
+ * @returns {Operation}
+ */
+export function uploadBlob(containerName, blobName, content, uploadOptions) {
```
Since every operation takes a containerName as the first argument, is this something that should go on `state.configuration`? Would a single job ever need to use two containers? Is a credential/API key linked to or more containers?

The API for all functions would be simplified if `containerName` could go on state.config.

bensonmiller replied:
updated to support this

josephjclark reviewed (commented):
Hello @bensonmiller thank you for such a well written and well-documented PR.

There is absolutely nothing wrong with your Javascript and frankly I think you should have a little more faith in yourself! It's all nice and clean and concise and my comments are more around high-level design or low-level patterns that OpenFn uses (and perhaps doesn't document so well).

I really like the idea of splintering out this "mini adaptor" to just cover part of Azure :ok_hand:  I wish we'd thought of that elsewhere.

I haven't tested anything yet. Regarding that, given that is adaptor is by you and for you, I am happy to take it on your word that it works. I can see that the API is fairly simple (and maybe will get simpler?) which again makes me feel like I don't need to test too extensively.

A temporary access key could be a good way to go but let's get through this review round first and then see where we are!

bensonmiller commented:
@josephjclark -- Thank you for this superb review. I've skimmed it and I love the feedback. I have a fairly busy day today, so it's unlikely that I'll have any updates to commit before tomorrow, but I'll work through your comments and incorporate at soonest.

josephjclark commented on packages/azureblobstorage/README.md:
````diff
@@ -0,0 +1,88 @@
+Create a file `job.js`, as shown below, to run with the OpenFn CLI.
+
+```js
+const data = state.data;
````
Oh, I forgot something really important! To make this code work we'll need to use `expandReferences` on the arguments to resolve the functions. I can't expand on that right now so this is really a reminder to me (check msgraph for a reference if I forgot to come back to you)

bensonmiller commented on packages/azureblobstorage/src/Adaptor.js:
```diff
@@ -0,0 +1,216 @@
+ * @param {string} blobName - Name of the blob to check.
+ * @returns {Operation}
+ */
+export function checkBlobExists(containerName, blobName) {
```
My main thought is that there may be circumstances when people don't want to clobber an existing blob. Given this, it probably makes sense to change the `uploadBlob` signature to let the user control the overwrite behavior  (`overwrite = true | false`) and eliminate this public function.

josephjclark replied:
Yep -  a flag on the upload looks like a better solution to me!

You'll need to decider whether to throw and kill the job, or to return a status to state or via callback. I presume that if the write fails the.jon writer will want to change the id or something, rather than abandon the job 🤔

bensonmiller replied:
I need to think more about this. For the moment, I lean towards killing the job and handling exceptions through a failure edge.

josephjclark replied:
I think the pattern you have now is great - let's use it as-is and see how it feels "in the wild".

mtuchi commented on packages/azureblobstorage/package.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "@openfn/language-azureblobstorage",
```
@josephjclark Azure have other storage services, [see Azure storage data services here]( https://learn.microsoft.com/en-us/azure/storage/common/storage-introduction?toc=/azure/storage/blobs/toc.json&bc=/azure/storage/blobs/breadcrumb/toc.json#azure-storage-data-services), Does this mean we have to create an adaptor for each storage service when they are needed ?

josephjclark replied:
Well not necessarily - but I'm wondering if it might be a good pattern to adopt. 

It will mean easier-to-understand adaptors, shorter docs, tighter APIs.

A big consideration is whether you want to use two services in a single job. If it's common to use blob storage and files in the same job/task, then it makes sense to consolidate.

And yes, rather than just an `azure-blobstorage`, it might make sense to have `azure-storage`.

Just thoughts for future adaptors!

bensonmiller commented on packages/azureblobstorage/src/Adaptor.js:
```diff
@@ -0,0 +1,216 @@
+  const { accountName, accountKey } = state.configuration;
+  const blobStorageUrl = `https://${accountName}.blob.core.windows.net`;
+  const sharedKeyCredential = new StorageSharedKeyCredential(accountName, accountKey);
+  state.client = new BlobServiceClient(blobStorageUrl, sharedKeyCredential);
```
updated

bensonmiller commented on packages/azureblobstorage/src/Adaptor.js:
```diff
@@ -0,0 +1,216 @@
+  return state;
+}
+
+// Destroy client
```
deleted

bensonmiller commented on packages/azureblobstorage/src/Adaptor.js:
```diff
@@ -0,0 +1,216 @@
+
+// A basic Azure BlobServiceClient client is attached to state.client
+// but almost all operations require a ContainerClient. This function
+// just keeps things a little DRYer.
```
resolved

bensonmiller commented on packages/azureblobstorage/src/Adaptor.js:
```diff
@@ -0,0 +1,216 @@
+ * @param {string} containerName - Name of the container to create.
+ * @returns {Operation}
+ */
+export function createContainer(containerName) {
```
implementation added

bensonmiller commented on packages/azureblobstorage/src/Adaptor.js:
```diff
@@ -0,0 +1,216 @@
+ * @param {string} blobName - Name of the blob to download.
+ * @returns {Operation}
+ */
+export function downloadBlobAsString(containerName, blobName) {
```
I updated this to support the `{as}` option, but didn't support `stream`. I might need an assist on implementing stream support.

bensonmiller commented on packages/azureblobstorage/assets/rect-512x200.png:
I've updated these images. I've been testing via the CLI. I'll look forward to seeing how it looks in Lightning.

josephjclark replied:
Thanks!

bensonmiller commented:
@josephjclark I've updated this PR to incorporate almost all of your comments. Notable changes include:

- Pulled the client out of `state` into a variable
- Eliminate `checkBlobExists()` function
- Unified `downloadBlob()` API with options to download as `string | json`
- Updated `uploadBlob()` function to support creation of container (boolean) or overwriting existing blobs (boolean) 
- Renamed adaptor to `azure-storage`
- Functions default to sourcing the container name from `state.configuration`, but the name can be overridden by a function argument if desired
- Added Azure Storage images to `assets/` directory
- Removed junk comments
- Adjusted log levels

josephjclark commented on packages/azure-storage/src/Adaptor.js:
```diff
@@ -0,0 +1,179 @@
+
+
+/**
+ * Download a blob from Azure Blob Storage as a string.
```
Description here needs a tweak

josephjclark commented on packages/azure-storage/src/Adaptor.js:
```diff
@@ -0,0 +1,179 @@
+ * @param {string} containerName - Name of the container containing the blob.
+ * @returns {Operation}
+ */
+export function downloadBlob(blobName, options = { as: 'string' }, containerName = null) {
```
stick container name in options

josephjclark commented on packages/azure-storage/src/Adaptor.js:
```diff
@@ -0,0 +1,179 @@
+  };
+}
+
+// TODO: Decide which functions to publish from @openfn/language-common
```
this todo ought to be removed

josephjclark commented on packages/azureblobstorage/CHANGELOG.md:
```diff
@@ -0,0 +1 @@
+# @openfn/language-azureblobstorage
```
We'll take care of this before releasing :)

bensonmiller commented on packages/azure-storage/README.md:
````diff
@@ -0,0 +1,92 @@
+Create a file `job.js`, as shown below, to run with the OpenFn CLI.
+
+```js
+const data = state.data;
````
@josephjclark -- Thank you for this comment. I've reworked my example per your guidance and I think I'm very close but I'm experiencing a problem with the way that your example uses the open function to return the name of the blob. For my experiment, I'm using the following two job definitions. The first one works. The second one throws an error when trying to instantiate the `blockBlobClient` (which is the first place where the `blobName` argument is used).

This works:
```
uploadBlob(
"some/path/myblobname.json",
  state.data,
  {
    blobHTTPHeaders: { blobContentType: 'application/json' }
  },
  { createContainer: true, containerName: 'thisone'}
);
```

This doesn't work
```
uploadBlob(
  () => { return "this/myblobname.json"; },
  state.data,
  {
    blobHTTPHeaders: { blobContentType: 'application/json' }
  },
  { createContainer: true, containerName: 'thisone'}
);
```

I run the job via the CLI. When it fails, it returns the following output. The `blobName.split is not a function` error is thrown when I try to instantiate the `blockBlobClient`. I'm assuming that function hasn't yet returned, so the `blockBlobClient` is trying to instantiate itself with a function argument that causes it to error.

```
➜  nhgh_workflows openfn uploadAzureSimpler.js -ma azure-storage -s tmp/state_for_azure_upload_simple.json -O  -v
[CLI] ♦ Versions:
         ▸ node.js                           18.17.1
         ▸ cli                               0.4.10
         ▸ runtime                           0.2.1
         ▸ compiler                          0.0.38
         ▸ @openfn/language-azure-storage    latest
[CLI] ✔ Loading adaptors from monorepo at /Users/benson/sandbox/newhorizons/openfn_adaptors
[CLI] ✔ Loaded state from tmp/state_for_azure_upload_simple.json
[CLI] ✔ Compiled from uploadAzureSimple.js
[R/T] ♦ Starting job job-1
[R/T] ✘ Failed job job-1 after 154ms
[R/T] ✘ TypeError: TypeError: blobName.split is not a function
[R/T] ✘ Check state.errors.job-1 for details.
[CLI] ✔ Result: 
[CLI] ♦ {
  "data": {
    "foo": "bar",
    "bif": "baz",
    "sunday": "monday",
    "answer": "42"
  },
  "errors": {
    "job-1": {
      "type": "TypeError",
      "jobId": "job-1",
      "message": "TypeError: blobName.split is not a function",
      "error": {
        "source": "runtime",
        "type": "RuntimeError",
        "severity": "fail",
        "subtype": "TypeError",
        "name": "RuntimeError",
        "message": "TypeError: blobName.split is not a function"
      }
    }
  }
}
[CLI] ⚠ Errors reported in 1 jobs
[CLI] ✔ Finished in **231ms**
```

I'd appreciate your thoughts on this. I'd be happy to jump on a GMT call if it would help. Thank you!

josephjclark replied:
Hi @bensonmiller - that's the `expandReferences` thing biting us.

Check out this comment: https://github.com/OpenFn/adaptors/pull/438#discussion_r1410435524

You need to expand the `blobName` reference and use the "resolved" value in the function.

bensonmiller replied:
Ah, yes! Thanks for the re-iteration. This should be all fixed up now.

bensonmiller commented:
@josephjclark I think (hope) that this is ready for a final look before prepping for release (`pnpm changeset`, etc.)
